### PR TITLE
fix crash when can't find killall command

### PR DESCRIPTION
**Pull request by nmyggg**
### Type of changes

_Put an `x` inside the [ ] that applies._

* [x] Bugfix
* [ ] New feature
* [ ] Translation
* [x] General refinement
* Other:

### Details

Please provide more details about changes if necessary.

some devices no killall command

### CODE REVIEW

**Comment by Mygod** [comment]
Does your ROM not have killall? What's your device?

**Comment by madeye** [comment]
@Mygod I think it's a good fix. There are still many AOSP ROMs don't have `killall`.

File: mobile/src/main/scala/com/github/shadowsocks/bg/Executable.scala [hunk 1]
```diff
@@ -18,10 +21,11 @@ object Executable {
...
-    val killer = new ProcessBuilder("killall" +: EXECUTABLES: _*).start()
-    if (!killer.waitFor(1, TimeUnit.SECONDS))
-      Log.w("killall", "%s didn't exit within 1s. Post-crash clean-up may have failed."
-        .formatLocal(Locale.ENGLISH, killer.toString))
+  def killAll(context: Context) {
+    val manager = context.getSystemService(Context.ACTIVITY_SERVICE).asInstanceOf[ActivityManager]
+
+    manager.getRunningAppProcesses.foreach(f => if (f.processName.equals(context.getPackageName + ":bg")) {
+      android.os.Process.killProcess(f.pid)
+    })
```
Comment (Mygod) [inline mobile/src/main/scala/com/github/shadowsocks/bg/Executable.scala]: Could this kill the bg Java process as well? That would be undesirable.

File: mobile/src/main/scala/com/github/shadowsocks/bg/Executable.scala [hunk 2]
```diff
@@ -18,10 +21,11 @@ object Executable {
 
   val EXECUTABLES = Array(SS_LOCAL, SS_TUNNEL, PDNSD, REDSOCKS, TUN2SOCKS, OVERTURE)
 
-  def killAll() {
-    val killer = new ProcessBuilder("killall" +: EXECUTABLES: _*).start()
-    if (!killer.waitFor(1, TimeUnit.SECONDS))
-      Log.w("killall", "%s didn't exit within 1s. Post-crash clean-up may have failed."
-        .formatLocal(Locale.ENGLISH, killer.toString))
+  def killAll(context: Context) {
+    val manager = context.getSystemService(Context.ACTIVITY_SERVICE).asInstanceOf[ActivityManager]
```
Comment (Mygod) [inline mobile/src/main/scala/com/github/shadowsocks/bg/Executable.scala]: I don't see a reason to use context here. Just use global app object defined in ShadowsocksApplication.

File: mobile/src/main/scala/com/github/shadowsocks/bg/Executable.scala (hunk 1 above)
Comment (madeye) [reply mobile/src/main/scala/com/github/shadowsocks/bg/Executable.scala]: Yes, we should avoid kill Java processes. Please kill only native processes.

Comment (nmyggg) [reply mobile/src/main/scala/com/github/shadowsocks/bg/Executable.scala]: ok it's on your call

**Comment by madeye** [comment]
BTW, I don't think we can list the native processes through Android APIs.

So, maybe it's still a good way to use `killall` here as a "best effort" crashing recovery approach.

**Comment by Mygod** [comment]
I just verified that `killall` does indeed exist on the Android emulator. It seems reasonable to assume that it exists for most devices. However a fallback would still be good. Let me look into this.

**Comment by Mygod** [comment]
Update: `killall` doesn't exist on Android 4.4 emulator.

**Comment by Mygod** [comment]
@madeye Yes `getRunningAppProcesses` only lists Java processes. Closing.

**Pull request closed without merging**